### PR TITLE
Soft-deprecate Axes.plot_date()

### LINKED ISSUE #18154: Deprecate plot_date()
*This is a writeup / proposal to serve as a basis for discussion in the dev call on 03 Aug 2020.*

`plot_date()` is essentially a thin wrapper around `plot()` but has different defaults for `fmt`. This causes confusion #17548, #18141.

### Context

- **Use case:** `plot_date()` dates back way before Matplotlib supported units. It's nowadays only necessary if you have plain numbers (in matplotlibs epoch). `datetime`s are directly handled in `plot()` via the unit system.
While there may be users for `plot_date()`, i suppose it's rather a niche application.

- **Naming:** The name is a bit misleading in that it suggests that you should use the function if you want to plot dates. This implicitly includes `datetime` / `date` which would be misleading. I'm suspecting that it may lead users with `datetime`/`date` objects on the wrong track. (`plot_as_date()` would have been more appropriate, but we're not gonna change that).

- **API design:** `plot_date()` is an ad-hoc aggregation of `plot()` and `axis_date()`
  ~~~
    def plot_date(self, x, y, fmt='o', tz=None, xdate=True, ydate=False, **kwargs):
        if xdate:
            self.xaxis_date(tz)
        if ydate:
            self.yaxis_date(tz)
        return self.plot(x, y, fmt, **kwargs)
  ~~~ 

### Recommendation
Given the above context, I propose to deprecate `plot_date()` in favor of `ax.plot()` and `ax.x/yaxis.axis_date()`.

At the same time, deprecate `ax.x/yaxis_date()` in favor of `ax.x/yaxis.axis_date()`. This reduces the API footprint of `Axes`. IMHO we don't need to have wrappers `x/yfoo()` wrappers for every `foo()` function on the axis. While common ones like `set_x/ylabel()` have a justification, it's fine in general to use `ax.xaxis.foo()`.

#### Alternatives:

- Keep as is - I wouldn't do that as it's confusing users.
- Change the default `fmt` to match `plot()` - Such a default change requires as much user action as the deprecation, but we keep all the unnecessary redundancy / imperfect naming.
- Only soft-deprecate (i.e. add a note to discourage the use but keep the function). This would be an option if we're strongly concerned with not forcing users to change their `plot_date()` code.

### PR DESCRIPTION
**Pull request by timhoffm**
## PR Summary

Closes #18154.

### CODE REVIEW

**Comment by timhoffm** [comment]
Yes, it's a little late here ...

I've made this a pending deprecation, which feels a bit better than a very long deprecation period.

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -1757,13 +1757,26 @@ def plot(self, *args, scalex=True, scaley=True, data=None, **kwargs):
...
                   **kwargs):
         """
         Plot data that contains dates.
 
+        .. note::
+
+            This method exists for historic reasons and will be deprecated in
+            the future.
+
+            - ``datetime`` data should directly be plotted using `~.Axes.plot`.
```
Comment (jklymak) [inline lib/matplotlib/axes/_axes.py:1774]: is ``datetime`` good enough, or should we also specify `np.datetime64`?

Comment (dstansby) [reply lib/matplotlib/axes/_axes.py:1774]: Maybe ``datetime``-like?

**Review by tacaswell — DISMISSED** [review]
:+1: in principle, but I think that this need a note in the API changes page.

**Review by dstansby — APPROVED** [review]

**Review by dstansby — CHANGES_REQUESTED** [review]
Woops, I approved before seeing the CI, that should probably be fixed! 👍 in theory though

**Comment by timhoffm** [comment]
`PendingDeprecationWarning` was not playing nicely with sphinx (warnings causing sphinx to fail). I suppose we'd have to include special handing for that.

Instead of a `PendingDeprecationWarning` I went back to only discouraging this code. In both cases we'd have to add a `DeprecationWarning` to later releases if we want to finally remove it. So there's no real benefit of the pending deprecation warning.

**Comment by timhoffm** [comment]
Macos test failure seems unrelated (subprocess timed out).

**Comment by jklymak** [comment]
@dstansby I think that CI is happier now...

**Review by dstansby — APPROVED** [review]

**Pull request merged**